io.openems.edge.ess.sma/src/io/openems/edge/ess/sma/stpxx3se/batteryinverter/BatteryInverterSmaStpSeImpl.java

```diff
@@ -0,0 +1,264 @@
+		this.config = config;
+		if (super.activate(context, config.id(), config.alias(), config.enabled(), config.modbusUnitId(), this.cm,
+				"Modbus", config.modbus_id(), READ_FROM_MODBUS_BLOCK)) {
+			return;
+		}
+		this._setGridMode(GridMode.ON_GRID);
+		this._setInitializing(true);
+	}
+
+	@Override
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+	}
+
+	@Override
+	protected void onSunSpecInitializationCompleted() {
+		this.logInfo(this.log, "SunSpec initialization finished. " + this.channels().size() + " Channels available.");
+
+		this.mapFirstPointToChannel(SymmetricBatteryInverter.ChannelId.ACTIVE_POWER,
+				ElementToChannelConverter.DIRECT_1_TO_1, DefaultSunSpecModel.S103.W);
+
+		this.mapFirstPointToChannel(SymmetricBatteryInverter.ChannelId.REACTIVE_POWER,
+				ElementToChannelConverter.DIRECT_1_TO_1, DefaultSunSpecModel.S103.V_AR);
+
+		this.mapFirstPointToChannel(SymmetricBatteryInverter.ChannelId.MAX_APPARENT_POWER,
+				ElementToChannelConverter.DIRECT_1_TO_1, DefaultSunSpecModel.S121.W_MAX);
+
+
+		this.getActivePowerChannel().onSetNextValue(val -> {		
+				this._setDcDischargePower(TypeUtils.subtract(val.get(), this.getDcPvPower()));
+		});
+
+		this._setInitializing(false);
+	}
+
+	@Override
+	public void run(Battery battery, int setActivePower, int setReactivePower) throws OpenemsNamedException {
+		this.calculateEnergy();
+		if (this.config.readOnly()) {
+			return;
+		}
+		if (battery instanceof SmaBattery sma) {
+			sma.applyPower(setActivePower, setReactivePower);
+			return;
+		}
+		this.logError(this.log, "Failed to run inverter. Battery is not SMA battery.");
+	}
+
+
+	/**
+	 * Calculate the Energy values from ActivePower.
+	 */
+	private void calculateEnergy() {
+		// AC
+		var activePower = this.getActivePower().get();
+		if (activePower == null) {
+			// Not available
+			this.calculateActiveChargeEnergy.update(null);
+			this.calculateActiveDischargeEnergy.update(null);
+		} else if (activePower > 0) {
+			// Buy-From-Inverter
+			this.calculateActiveChargeEnergy.update(0);
+			this.calculateActiveDischargeEnergy.update(activePower);
+		} else {
+			// Sell-To-Inverter
+			this.calculateActiveChargeEnergy.update(activePower * -1);
+			this.calculateActiveDischargeEnergy.update(0);
+		}
+		// DC
+		var dcPower = this.getDcDischargePower().get();
+		if (dcPower == null) {
+			// Not available
+			this.calculateDcChargeEnergy.update(null);
+			this.calculateDcDischargeEnergy.update(null);
+		} else if (dcPower > 0) {
+			// Discharging battery
+			this.calculateDcChargeEnergy.update(0);
+			this.calculateDcDischargeEnergy.update(dcPower);
+		} else {
+			// Charging battery
+			this.calculateDcChargeEnergy.update(dcPower * -1);
+			this.calculateDcDischargeEnergy.update(0);
+		}
+	}
+
+	@Override
+	public Integer getSurplusPower() {
+		// TODO Implement properly
+		return null;
+	}
+
+	@Override
+	public String debugLog() {
+		return "|L:" + this.getActivePower().asString(); //
```

Comment:
You could consider using the method 
`Battery.generateDebugLog(this, this.stateMachine);`
here